### PR TITLE
feat(java): support jar/war/ear

### PR DESCRIPTION
**Pull request by knqyf263**
## Overview
Support Java scanning 🎉 

```
$ trivy image --vuln-type library jenkins:2.60.3-alpine

usr/share/jenkins/jenkins.war
=============================
Total: 7 (UNKNOWN: 0, LOW: 0, MEDIUM: 2, HIGH: 4, CRITICAL: 1)

+---------------------------------------+------------------+----------+-------------------+---------------+-----------------------------------------+
|                LIBRARY                | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |                  TITLE                  |
+---------------------------------------+------------------+----------+-------------------+---------------+-----------------------------------------+
| com.google.guava:guava                | CVE-2018-10237   | MEDIUM   | 11.0.1            | 24.1.1        | guava: Unbounded memory                 |
|                                       |                  |          |                   |               | allocation in AtomicDoubleArray         |
|                                       |                  |          |                   |               | and CompoundOrdering classes            |
|                                       |                  |          |                   |               | allow remote attackers...               |
|                                       |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2018-10237   |
+---------------------------------------+------------------+----------+-------------------+---------------+-----------------------------------------+
| commons-beanutils:commons-beanutils   | CVE-2014-0114    | HIGH     | 1.8.3             | 1.9.2         | Apache Struts 1: Class Loader           |
|                                       |                  |          |                   |               | manipulation via request parameters     |
|                                       |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2014-0114    |
+                                       +------------------+          +                   +---------------+-----------------------------------------+
|                                       | CVE-2019-10086   |          |                   | 1.9.4         | apache-commons-beanutils: does          |
|                                       |                  |          |                   |               | not suppresses the class property       |
|                                       |                  |          |                   |               | in PropertyUtilsBean by default         |
|                                       |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2019-10086   |
+---------------------------------------+------------------+----------+-------------------+---------------+-----------------------------------------+
| commons-fileupload:commons-fileupload | CVE-2016-1000031 | CRITICAL | 1.3.1-jenkins-1   | 1.3.3         | Apache Commons FileUpload:              |
|                                       |                  |          |                   |               | DiskFileItem file manipulation          |
|                                       |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2016-1000031 |
+                                       +------------------+----------+                   +---------------+-----------------------------------------+
|                                       | CVE-2014-0050    | HIGH     |                   | 1.3.1         | apache-commons-fileupload: denial       |
|                                       |                  |          |                   |               | of service due to too-small buffer      |
|                                       |                  |          |                   |               | size used by MultipartStream...         |
|                                       |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2014-0050    |
+                                       +------------------+          +                   +---------------+-----------------------------------------+
|                                       | CVE-2016-3092    |          |                   | 1.3.2         | tomcat: Usage of vulnerable             |
|                                       |                  |          |                   |               | FileUpload package can result           |
|                                       |                  |          |                   |               | in denial of service...                 |
|                                       |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2016-3092    |
+---------------------------------------+------------------+----------+-------------------+---------------+-----------------------------------------+
| org.apache.commons:commons-compress   | CVE-2018-11771   | MEDIUM   |              1.10 |          1.18 | apache-commons-compress:                |
|                                       |                  |          |                   |               | ZipArchiveInputStream.read()            |
|                                       |                  |          |                   |               | fails to identify correct EOF           |
|                                       |                  |          |                   |               | allowing for DoS via crafted...         |
|                                       |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2018-11771   |
+---------------------------------------+------------------+----------+-------------------+---------------+-----------------------------------------+
```

## Blocker
- https://github.com/aquasecurity/trivy-db/pull/105
- https://github.com/aquasecurity/go-dep-parser/pull/16
- https://github.com/aquasecurity/fanal/pull/152

### CODE REVIEW

**Comment by czunker** [comment]
Nice to see this feature. Thank you very much.

**Comment by czunker** [comment]
I gave it a try with a container build with Ubuntu 20.04 + openjdk 11 + maven 3.
The scanner finds this:
```
usr/share/java/guava.jar
========================
Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 1, HIGH: 0, CRITICAL: 0)

+------------------------+------------------+----------+-------------------+---------------+---------------------------------------+
|        LIBRARY         | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |                 TITLE                 |
+------------------------+------------------+----------+-------------------+---------------+---------------------------------------+
| com.google.guava:guava | CVE-2018-10237   | MEDIUM   |              19.0 | 24.1.1        | guava: Unbounded memory               |
|                        |                  |          |                   |               | allocation in AtomicDoubleArray       |
|                        |                  |          |                   |               | and CompoundOrdering classes          |
|                        |                  |          |                   |               | allow remote attackers...             |
|                        |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2018-10237 |
+------------------------+------------------+----------+-------------------+---------------+---------------------------------------+
```

It also does detect this jar, but with no vulnerabilities:
```
usr/share/java/wagon-http-shaded-3.3.4.jar
==========================================
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
```
Checking with another scanner, it seems this jar contains the lib `org.apache.httpcomponents:httpclient:4.5.11` which has a known vulnerability (CVE-2020-13956).

Taking a look at the snyk Vulnerability DB, it knows about this CVE:
https://snyk.io/vuln/search?q=CVE-2020-13956&type=maven

So, I'm not sure why this CVE does not show up in the scan.

**Comment by knqyf263** [comment]
Trivy depends on GitHub Security Advisories (GHSA) for now and it looks like GHSA doesn't have CVE-2020-13956.
https://github.com/advisories?query=CVE-2020-13956

We recently found another public database including CVE-2020-13956. After supporting GitLab Advisory Database, Trivy will be able to detect it.
https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/blob/master/maven/org.apache.httpcomponents/httpclient/CVE-2020-13956.yml

**Comment by czunker** [comment]
Thanks for the clarification, missed the piece on GitHub Security Advisories (GHSA).
From the issue, I assume snyk.

File: pkg/detector/library/driver.go [hunk 1]
```diff
@@ -57,17 +51,17 @@ type Driver struct {
 	advisories []advisory
 }
 
-// NewDriver is the factory method from drier
-func NewDriver(advisories ...advisory) Driver {
+// Aggregate aggregates drivers
+func Aggregate(advisories ...advisory) Driver {
```
Comment (masahiro331) [inline pkg/detector/library/driver.go:55]: Aggregate return driver, but the driver has not ecosystem. 
driver.Type() method return empty.

**Comment by masahiro331** [comment]
Add DataSource for Java.
https://github.com/aquasecurity/trivy#data-sources

GHSA maven url: https://github.com/advisories?query=ecosystem%3Amaven

**Comment by masahiro331** [comment]
:shipit:

**Review by masahiro331 — APPROVED** [review]
LGTM :shipit:

**Comment by knqyf263** [comment]
Note that we should support GitLab Advisory Database to detect more vulnerabilities.

**Comment by bitsofinfo** [comment]
eta?

**Comment by knqyf263** [comment]
To improve accuracy, we're waiting for GitLab.
https://gitlab.com/gitlab-org/gitlab/-/issues/321589

**Comment by fawind** [comment]
Very excited about this! I found it a bit hard to follow the current status of the Gitlab Vuln DB support. Did this get resolved and is there an ETA for the next Trivy release including this feature? Thanks!

**Comment by malanvaneck** [comment]
@fawind, me and you both! Being excited!

**Comment by prabhushan** [comment]
Hi,.. So Trivy is going to support Java too? this is really great.. Which version of Trivy is this supported 0.16?

**Comment by prabhushan** [comment]
I tried today.. with 0.16 and this is what it is showing
`trivy image --vuln-type library jenkins:2.60.3-alpine
2021-04-23T09:09:09.137-0400	INFO	Trivy skips scanning programming language libraries because no supported file was detected`

**Comment by jeanluclariviere** [comment]
@prabhushan although this change was merged to main, it hasn't been released yet. As knqyf263 stated above, they are waiting on   https://gitlab.com/gitlab-org/gitlab/-/issues/321589. If you want to test this, you should clone the main branch and build the binary yourself; I've done this and it's working nicely.

**Comment by prabhushan** [comment]
Thank you @jeanluclariviere

**Comment by Xaseron** [comment]
Looks like this feature is not compatible with the air gapped mode

**Comment by jeanluclariviere** [comment]
@Xaseron I haven't had any issues with air-gap. I've been using https://github.com/sighupio/trivy-offline and as soon as the new version was released, my jars, wars etc started getting scanned immediately.

This project uses the [official air-gap documentation](https://github.com/aquasecurity/trivy/blob/main/docs/air-gap.md) and rebuilds an image daily to be used.

**Comment by Xaseron** [comment]
We are using artifactory to proxy the offline db.
I don't know why it is trying to access maven.org
```
[2021-05-05T08:22:47.007Z] + mkdir -p trivy/db
[2021-05-05T08:22:47.008Z] + wget https://artifactory.example.com/github-generic/aquasecurity/trivy-db/releases/latest/download/trivy-offline.db.tgz -O trivy/db/trivy-offline.db.tgz
[2021-05-05T08:22:47.008Z] Connecting to artifactory.example.com (x.x.x.x:443)
[2021-05-05T08:22:47.963Z] saving to 'trivy/db/trivy-offline.db.tgz'
[2021-05-05T08:22:48.230Z] trivy-offline.db.tgz  24% |*******                         | 5431k  0:00:03 ETA
[2021-05-05T08:22:49.197Z] trivy-offline.db.tgz 100% |********************************| 21.6M  0:00:00 ETA
[2021-05-05T08:22:49.197Z] 'trivy/db/trivy-offline.db.tgz' saved
[2021-05-05T08:22:49.197Z] + tar -xf trivy/db/trivy-offline.db.tgz -C trivy/db
[2021-05-05T08:22:52.568Z] + trivy --cache-dir ./trivy --timeout 10m0s --debug --skip-update --format template --template @/contrib/junit.tpl --output junit-trivy-report-low.xml --severity UNKNOWN,LOW,MEDIUM docker-dev-local.artifactory.example.com/event-import-service:1.5.8-SNAPSHOT
[2021-05-05T08:22:52.568Z] 2021-05-05T08:22:52.262Z	[35mDEBUG[0m	Severities: UNKNOWN,LOW,MEDIUM
[2021-05-05T08:22:52.568Z] 2021-05-05T08:22:52.300Z	[35mDEBUG[0m	cache dir:  ./trivy
[2021-05-05T08:22:52.568Z] 2021-05-05T08:22:52.301Z	[35mDEBUG[0m	DB Schema: 1, Type: 1, UpdatedAt: 2021-05-05 00:41:20.251299122 +0000 UTC, NextUpdate: 2021-05-05 12:41:20.251298422 +0000 UTC, DownloadedAt: 0001-01-01 00:00:00 +0000 UTC
[2021-05-05T08:22:52.568Z] 2021-05-05T08:22:52.301Z	[35mDEBUG[0m	Vulnerability type:  [os library]
[2021-05-05T08:23:00.839Z] 2021-05-05T08:23:00.725Z	[35mDEBUG[0m	Parsing Java artifacts...	{"file": "opt/java/openjdk/lib/jrt-fs.jar"}
[2021-05-05T08:23:01.866Z] 2021-05-05T08:23:01.680Z	[35mDEBUG[0m	request failed	{"error": "Get \"http://search.maven.org/solrsearch/select?q=1%3A%228e6365acaacea5ff57a744530cde01dd97e0b753%22&rows=1&wt=json\": dial tcp 3.214.218.146:80: connect: connection refused", "method": "GET", "url": "http://search.maven.org/solrsearch/select?q=1%3A%228e6365acaacea5ff57a744530cde01dd97e0b753%22&rows=1&wt=json"}
[2021-05-05T08:23:01.866Z] 2021-05-05T08:23:01.680Z	[35mDEBUG[0m	retrying request	{"request": "GET http://search.maven.org/solrsearch/select?q=1%3A%228e6365acaacea5ff57a744530cde01dd97e0b753%22&rows=1&wt=json", "timeout": "20s", "remaining": 5}
```

**Pull request merged**